### PR TITLE
fix(core): throw an error when no elements found by DebugElement.query

### PR DESCRIPTION
**Pull request by lacolaco**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #22449

Supersedes #32365

Currently, DebugElement.query 's return type is only DebugElement but actually, it returns DebugElement | null.

## What is the new behavior?

It returns only non-null `DebugElement`. If no children found, it will throw an error.

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

BREAKING CHANGE: Despite its typing is returning non-null `DebugElement`, `DebugElement.query` returned null when that predicate didn't match any children,. Currently, it throws an error to align the typing.

This breaking change is not too impactful because if this change causes an error, users should have already raised a run-time exception due to an unexpected null.

## Other information

cc/ @petebacondarwin

### CODE REVIEW

**Comment by petebacondarwin** [comment]
Oh man! This creates just as much churn in the specs as the other version. 😢 
I am not sure if either approach is better from a developer experience point of view.
And I can imagine how it would upset a lot of developers to suddenly find many of their tests either failing to compile or failing at runtime when they just upgrade Angular.

@AndrewKushnir also pointed out (offline) that this PR moves the API away from the standard `document.querySelector()` which also can return `null`.

It might be that leaving the typings as they are, while not ideal, would be the best option...

What is your view on this @lacolaco?

**Comment by lacolaco** [comment]
@petebacondarwin After implementing both versions, now I think the old `DebugElement | null` version is better because of as Andrew says. The word `query` is strongly suggestive of the `querySelector` API. Developers may not expect it will throw an error. 

Changing the `DebugElement.query` API is difficult because many test code including Angular itself uses it heavily and they expects it returns `null` as a common practice...

How about adding a new type-safe API to `DebugElement` for deprecating the old one, like `TestBed.get`? For example, `find(predicate)` and `findAll(predicate)`.

**Comment by crisbeto** [comment]
For what it's worth, we may be able to write an automated migration for this in v12. We had one breaking change in v11 that made a type nullable (https://github.com/angular/angular/pull/39009).

**Comment by AndrewKushnir** [comment]
> How about adding a new type-safe API to DebugElement for deprecating the old one, like TestBed.get? For example, find(predicate): DebugElement | null and findAll(predicate): ReadonlyArray<DebugElement>.

Just a quick comment on that: I think that introducing new APIs (find/findAll or may be directly mimic querySelector/querySelectorAll) is a good way to avoid breaking changes, but the downside is that deprecating and later removing old APIs at some point would make a lot of articles/guides/materials (created by the community) outdated, given that the `query` function usage is pretty significant. As @crisbeto mentioned, as a possible option, we might consider adding a migration to help transition application code.

Given that I'm still not 100% sure what is the right approach for this change and I believe more discussion is needed...

**Review by gbumanzordev — APPROVED** [review]

File: packages/platform-browser/testing/src/matchers.ts [hunk 1]
```diff
@@ -281,10 +281,12 @@ _global.beforeEach(function() {
...
-          const found = !!actualFixture.debugElement.query(By.directive(expectedComponentType));
-          return found ?
-              {pass: true} :
-              {pass: false, message: msgFn(`Expected ${expectedComponentType.name} to show`)};
+          try {
+            actualFixture.debugElement.query(By.directive(expectedComponentType));
+            return {pass: true};
+          } catch {
+            return {pass: false, message: msgFn(`Expected ${expectedComponentType.name} to show`)};
+          }
```
Comment (gkalpak) [inline packages/platform-browser/testing/src/matchers.ts:289]: Using `try`/`catch` for control flow isn't great. Could we use `queryAll()` and check for `.length > 0` instead?

File: aio/src/app/shared/select/select.component.spec.ts [hunk 2]
```diff
@@ -180,8 +180,8 @@ function getButtonSymbol(): HTMLElement | null {
 }
 
 function getOptionContainer(): HTMLUListElement|null {
-  const de = element.query(By.css('ul'));
-  return de && de.nativeElement;
+  const de = element.queryAll(By.css('ul'));
```
Comment (gkalpak) [inline aio/src/app/shared/select/select.component.spec.ts:183]: Nit: Consider using `element.queryAll(By.css('ul'))[0] ?? null` for consistency with other places.

**Review by gkalpak — COMMENTED** [review]
Drive-by review 🚗

**Comment by lacolaco** [comment]
I want to wait for the decision from the Angular team. The current typing for `DebugElement` has a problem, but solving it is not easy. 
By the way, should I write a design doc?

**Comment by atscott** [comment]
I ran a quick test to see how breaking the error throwing would be and found that it would also be pretty break-y and need a migration. There are plenty of places developers have `expect(queryResult).toBeNull()`, as they've discovered the true type. 

With that, I think the best thing to do here would be to align with the `document.querySelector` functionality and write a migration to add non-null assertions.

This change will need a migration, as @crisbeto mentioned (example: https://github.com/angular/angular/pull/39009). I'm going to close this but please do open a new PR when you've updated the approach to include `null` in the return type and have a migration with it. The migration should update all places that use `query` to have the non-null assertion. Even if the test later has an `expect(result).toBeNull()`, the migration's aim should be to adjust code to be equivalent to the current state.

**Pull request closed without merging**